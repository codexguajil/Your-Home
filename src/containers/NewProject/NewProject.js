import React from 'react';
import { Button, Modal, Form, Input, Steps, message, Menu, Dropdown, Icon, Card, Table, Divider, Tag } from 'antd';

const Step = Steps.Step;

const ProjectForm = Form.create({ name: 'form_in_modal' })(
  class extends React.Component {

    renderSteps = (getFieldDecorator, current, next, prev, rooms, addRoom, onCreate, name, addName, deleteItem) => {

      const onClick = function ({ key }) {
        message.info(`Added a ${key}`);
        addRoom(key, Date.now())
      };

      const menu = (
        <Menu onClick={onClick}>
          <Menu.Item key="Kitchen">Add Kitchen</Menu.Item>
          <Menu.Item key="Bedroom">Add Bedroom</Menu.Item>
          <Menu.Item key="Bathroom">Add Bathroom</Menu.Item>
          <Menu.Item key="Living Room">Add Living Room</Menu.Item>
          <Menu.Item key="Outdoors">Add Outdoors</Menu.Item>
        </Menu>
      );

      const dataSource = [
        {
          key: '1',
          name: `${name}`,
          rooms: `${rooms.length}`,
        },
      ];

      const columns = [
        {
          title: 'Name',
          dataIndex: 'name',
          key: 'name',
        },
        {
          title: 'Rooms',
          dataIndex: 'rooms',
          key: 'rooms',
        }
      ]

      const steps = [
        {
          title: 'Project Name',
          content: <Form layout="vertical">
                      <Form.Item label="Name">
                        {getFieldDecorator('name', {
                          rules: [{ required: true, message: 'Please input the name of the project!' }],
                        })(<Input onChange={addName}/>)}
                      </Form.Item>
                      <Form.Item label="Description">
                        {getFieldDecorator('description')(<Input type="textarea" />)}
                      </Form.Item>
                    </Form>
        },
        {
          title: 'Rooms',
          content: <div>
                    <Card title="Card title" className="card">
                      {
                        rooms &&
                        rooms.map(room => <div className="room-item">
                                              <p>{room.name}</p>
                                              <button name={room.id} 
                                                      onClick={deleteItem}>
                                                      x
                                               </button>
                                          </div>)
                      }
                    </Card>
                    <Dropdown overlay={menu}>
                      <a className="ant-dropdown-link" href="#">
                        Select a Room Type <Icon type="down" />
                      </a>
                    </Dropdown>
                   </div>
        },
        {
          title: 'Create Project',
          content: <div>
                    <Card title="Review Project" className="card">
                    <Table pagination={false} dataSource={dataSource} columns={columns} />
                    </Card>
                   </div>
        }
      ]

      return (
        <div>
          <Steps current={current}>
            {steps.map(item => (
              <Step key={item.title} title={item.title} />
            ))}
          </Steps>
          <div className="steps-content">{steps[current].content}</div>
          <div className="steps-action">
            {current > 0 && (
              <Button style={{ marginLeft: 8 }} onClick={() => prev()}>
                Previous
              </Button>
            )}
            {current < steps.length - 1 && (
              <Button type="primary" onClick={() => next()}>
                Next
              </Button>
            )}
          </div>
        </div>
      )
    }

    render() {
      const { visible, onCancel, onCreate, form, current, next, prev, rooms, addRoom, onSubmit, name, addName, deleteItem } = this.props;
      const { getFieldDecorator } = form;
      return (
        <Modal
          visible={visible}
          title="Create a new project"
          okText="Create"
          onCancel={onCancel}
          onOk={onSubmit}
        >
        {this.renderSteps(getFieldDecorator, current, next, prev, rooms, addRoom, onCreate, name, addName, deleteItem)}
        </Modal>
      );
    }
  },
);

class NewProject extends React.Component {
  state = {
    visible: false,
    current: 0,
    rooms: [],
    name: ''
  };

  next = () => {
    if(this.state.current === 0) {
      this.handleCreate()
    } else {
      const current = this.state.current + 1;
      this.setState({ current });
    }
  }

  addName = (e) => {
    console.log('working')
    this.setState({name: e.target.value})
  }

  deleteRoomItem = (e) => {
    let newRooms = []
    this.state.rooms.forEach(room => {
      if(room.id != e.target.name) {
        newRooms.push(room)
      }
    })
    this.setState({rooms: newRooms})
  }

  nextField = () => {
    const current = this.state.current + 1;
    this.setState({ current });
  }

  prev = () => {
    const current = this.state.current - 1;
    this.setState({ current });
  }

  submit = () => {
    console.log(this.state)
    this.setState({ visible: false });
  }

  addRoom = (room, id) => {
    this.setState({ rooms: [...this.state.rooms, {name: room, id}]})
  }

  showModal = () => {
    this.setState({ visible: true });
  };

  handleCancel = () => {
    this.setState({ visible: false });
  };

  render() {
    return (
      <div id='add'>
        <Button 
          type="primary" 
          onClick={this.showModal}>
            New Project
        </Button>
        <ProjectForm
          wrappedComponentRef={this.saveFormRef}
          visible={this.state.visible}
          onCancel={this.handleCancel}
          // onCreate={this.handleCreate}
          current={this.state.current}
          next={this.nextField}
          prev={this.prev}
          rooms={this.state.rooms}
          addRoom={this.addRoom}
          onSubmit={this.submit}
          name={this.state.name}
          addName={this.addName}
          deleteItem={this.deleteRoomItem}
        />
      </div>
    );
  }
}

export default NewProject;