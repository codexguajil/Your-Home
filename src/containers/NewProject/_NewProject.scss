// #add-project {
//   position: absolute;
//   right: 44px;
//   top: 42px;
// }

#add {
  display: inline;
  position: absolute;
  right: 25px;
  top: 25px;
}