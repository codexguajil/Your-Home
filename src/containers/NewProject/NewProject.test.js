import React from 'react';
import { shallow } from 'enzyme';
import NewProject from './NewProject';
// import ProjectForm from './NewProject';

describe('App', () => {
  let wrapper;
  
  beforeEach(() => {
    wrapper = shallow(<NewProject/>)
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it.skip('should setState on nextField', () => {
    wrapper.setState('current').toBe(0);
    wrapper.instance().nextField();
    expect(wrapper.state('current')).toEqual(1);
  });

  it.skip('should setState on prev', () => {
    wrapper.setState('current').toBe(1);
    wrapper.instance().prev();
    expect(wrapper.state('current')).toEqual(0);
  });

  it('should setState on submit', () => {
    wrapper.instance().submit();
    expect(wrapper.state('visible')).toBe(false);
  });

  it('should setState on addRoom', () => {
    wrapper.instance().addRoom('Kitchen');
    expect(wrapper.state('rooms')).toEqual(['Kitchen']);
  });

  it('should setState on showModal', () => {
    wrapper.instance().showModal();
    expect(wrapper.state('visible')).toBe(true);
  });

  it('should setState on handleCancel', () => {
    wrapper.instance().handleCancel();
    expect(wrapper.state('visible')).toBe(false);
  });
});