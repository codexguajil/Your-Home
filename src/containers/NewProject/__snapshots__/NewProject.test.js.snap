// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  id="add"
>
  <Button
    block={false}
    ghost={false}
    htmlType="button"
    loading={false}
    onClick={[Function]}
    type="primary"
  >
    New Project
  </Button>
  <Form(_temp)
    addName={[Function]}
    addRoom={[Function]}
    current={0}
    name=""
    next={[Function]}
    onCancel={[Function]}
    onSubmit={[Function]}
    prev={[Function]}
    rooms={Array []}
    visible={false}
  />
</div>
`;

exports[`App should match snapshot 2`] = `
<_temp
  form={
    Object {
      "getFieldDecorator": [Function],
      "getFieldError": [Function],
      "getFieldInstance": [Function],
      "getFieldProps": [Function],
      "getFieldValue": [Function],
      "getFieldsError": [Function],
      "getFieldsValue": [Function],
      "isFieldTouched": [Function],
      "isFieldValidating": [Function],
      "isFieldsTouched": [Function],
      "isFieldsValidating": [Function],
      "isSubmitting": [Function],
      "resetFields": [Function],
      "setFields": [Function],
      "setFieldsInitialValue": [Function],
      "setFieldsValue": [Function],
      "submit": [Function],
      "validateFields": [Function],
      "validateFieldsAndScroll": [Function],
    }
  }
/>
`;
